### PR TITLE
Use double precision in threaded calculation of linear tree coefficients (fixes #5226)

### PR DESCRIPTION
**Pull request by btrotta**
Fixes https://github.com/microsoft/LightGBM/issues/5266

When calculating the linear tree coefficients, we need to calculate some matrix products. This calculation is multi-threaded for efficiency. However, floating-point addition can give slightly different results depending on the order terms are added, and this was causing the calculated coefficients to vary depending on the number of threads. I've resolved this by making these matrices double precision instead of single, so the inaccuracies are less significant. This will not use much additional memory, since the size of these matrices is only `O(num_features ^ 2)`.

I also added a preprocessor directive to make Eigen calls single-threaded (since Eigen is always called inside a for-loop that is already parallelized.

### CODE REVIEW

File: CMakeLists.txt
Comment (jameslamb) [inline CMakeLists.txt:112]: Can you please add this to the flags used by the R package as well?

https://github.com/microsoft/LightGBM/blob/44fe591a60d7427d64997c37b22768a92c97c47b/R-package/configure.ac#L38

https://github.com/microsoft/LightGBM/blob/44fe591a60d7427d64997c37b22768a92c97c47b/R-package/configure.win#L22

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks so much for fixing this! I left a few initial suggestions for your consideration.

File: tests/python_package_test/test_engine.py
Comment (btrotta) [reply tests/python_package_test/test_engine.py]: I've changed it to use some randomly generated data (I was unable to reproduce the failure with the breast cancer dataset).

Comment (jameslamb) [reply tests/python_package_test/test_engine.py]: Ah ok, I didn't realize that it could be dataset-specific. Sorry for creating extra work for you! I guess we could have also avoided "test file importing another test file" by moving `FileLoader` to `utils.py`. That would probably good to do *anyway* (in a separate PR).

Anyway, if the new randomly-generated data is sufficient to reproduce the underlying issue fixed by this PR, I'm good with it!

File: CMakeLists.txt
Comment (jameslamb) [reply CMakeLists.txt:112]: For non-Windows, we have a comment-triggered CI job that will update `configure` based on changes to `configure.ac`. I'll trigger that job right now.

For `configure.win`, nothing needs to be regenerated. `configure.win` is executed directly, instead of being used as a template.

These things are documented at https://github.com/microsoft/LightGBM/tree/master/R-package#changing-the-cran-package but that README is fairly large so it's easy to miss.

**Comment by jameslamb** [comment]
hmmm the run-configure job failed ([build link](https://github.com/microsoft/LightGBM/runs/7323912286?check_suite_focus=true))

```text
Error: fatal: couldn't find remote ref refs/heads/linear-threading
```

I'll try one more run, and if it still fails I'll look into this later. Sorry for the disruption 😭

**Comment by jameslamb** [comment]
I see the issue! I believe that updating `R-package/configure` on a PR from a fork doesn't currently work today. I've documented that in #5371 and will work on a fix, but it doesn't need to block this PR.

@btrotta I just regenerated `configure` locally using the dockerized steps mentioned at https://github.com/microsoft/LightGBM/tree/master/R-package#changing-the-cran-package.  I tried to push those changes to your branch (which I thought I could do as a maintainer here), but unfortunately I got a "permission denied".

Thankfully, autoconf only generated a change on one line.

Can you please change this line

https://github.com/btrotta/LightGBM/blob/32f564c26017fbae7c78663046dffd7ca0eec177/R-package/configure#L1716

to

```shell
LGB_CPPFLAGS="${LGB_CPPFLAGS} -DEIGEN_MPL2_ONLY -DEIGEN_DONT_PARALLELIZE"
```

Sorry again for the disruption.

**Comment by jameslamb** [comment]
On other thing... as a maintainer you have permissions to push branches directly to LightGBM instead of using your fork. I recommend doing that in the future.

**Comment by btrotta** [comment]
@jameslamb Thanks for your help, I've updated `R-package/configure` now

**Review by guolinke — APPROVED** [review]
Thank you!

**Comment by btrotta** [comment]
@jameslamb If you're happy with the changes, could you please approve?

**Review by StrikerRUS — APPROVED** [review]
Thanks a lot for the fix!

**Comment by jameslamb** [comment]
> If you're happy with the changes, could you please approve?

I'm leaving my "request changes" review and not re-reviewing until the issue that has been LightGBM's CI for the last few days (https://github.com/microsoft/LightGBM/pull/5362#issuecomment-1192780327) is fixed. I don't want this to be merged until the CI is fixed and we run it over these changes one more time.

Sorry for the delay. Hopefully that issue will be fixed soon.

**Comment by btrotta** [comment]
@jameslamb Ok, no problem.

**Comment by jameslamb** [comment]
Ok @btrotta our CI concerns have been resolved (#5388).

Since I don't have access to push to your fork...could you please merge latest `master` into this branch? Once you do that and CI passes, I'll merge this fix.

Sorry for the delay, and thanks for your patience.

**Comment by shiyu1994** [comment]
Maybe we can close and reopen this PR to rerun the ci? Since the ci tests are run with the merged version automatically.

**Comment by jameslamb** [comment]
> Maybe we can close and reopen this PR to rerun the ci? Since the ci tests are run with the merged version automatically.

I personally prefer to merge the target branch (in this repo, `master`) into PRs directly as a way to trigger CI, instead of just closing and re-opening, in situations like this one where the PR has changes to a central part of the codebase and when there are several other active PRs touching other related parts of the codebase.

To avoid this situation:

1. `branch-1` and `branch-2` are created off of `master` at the same time
2. changes made on `branch-1` are proposed as PR-1
3. changes made on `branch-2` are proposed as PR-2
4. CI passes on both PR-1 and PR-2
5. PR-1 is merged, then PR-2 is merged **without re-running CI on PR-2**
6. `master` is broken because the changes in PR-1 and PR-2 are incompatible (for example, PR-2 removes a function that was referenced by the code in PR-1)

If CI passes for this particular PR I'm ok with merging it to keep making progress on all the other PRs, but please consider that for the future.

**Comment by shiyu1994** [comment]
@jameslamb Thank you. I will follow this rule in the future.

**Comment by shiyu1994** [comment]
It seems that all ci tests are passed. Maybe we can cancel the change request now?

**Comment by jameslamb** [comment]
> Maybe we can cancel the change request now?

oh yep, sorry about that! Meant to come back and approve yesterday.

**Review by jameslamb — APPROVED** [review]
🎉 thanks again for the help @btrotta !

**Comment by jameslamb** [comment]
> I will follow this rule in the future.

No problem! I've found it helpful. There is definitely a trade-off there. Especially since our CUDA CI jobs take 1-2 hours to run and we can only have a single job running at a time across the whole repo, extra CI runs can slow down development in the whole project.

My approach is:

* if I need to re-trigger CI anyway, merge latest `master` and use that as a way to re-trigger it
* if a PR recently passed CI and has very small, non-functional changes such as documentation fixes, just merge it without rebuilding

So, for example, once #5384 builds successfully, I'll just merge it even though this PR was merged to `master` after that build started. Since #5384 just contains documentation changes.

**Pull request merged**
